### DISCUSSION TITLE
Reconstructing an Undersampled Spiral

### DISCUSSION THREAD

**Discussion by nathanjbeaumont** [JuliaHealth/KomaMRI.jl#506] [Q&A]
Hello everyone,
I have an undersampled spiral sequence that I want to reconstruct but I don't know how to go about it. The simulation signal looks as expected:
![image](https://github.com/user-attachments/assets/197f0280-1fc1-4680-86ec-18e3e7e6af11)
![image](https://github.com/user-attachments/assets/1acbcb35-b34d-46db-8ce2-8ea14b604a23)

But reconstruction is not handling the under sampling:
![image](https://github.com/user-attachments/assets/17a0258a-ebff-48d7-aaf6-d7e5bdd45382)

I'm expecting to see an image of similar quality to figure 7 here https://onlinelibrary.wiley.com/doi/10.1002/mrm.25559:
![image](https://github.com/user-attachments/assets/d221b52c-0aeb-41da-8e40-9ed068dc9629)

What parameters do I need to set in the reconstruction function to do this?

Here's the code:
```julia
using KomaMRICore, KomaMRIPlots, KomaMRI, CUDA
obj = brain_phantom2D()
plot_phantom_map(obj, :ρ)
seq = read_seq("spiralNormal.seq")
plot_seq(seq)
sys = Scanner() # default hardware definition
raw = simulate(obj, seq, sys)
p4 = plot_signal(raw)
acq = AcquisitionData(raw)
reconParams = Dict{Symbol,Any}(:reconSize=>(255, 255))
image = reconstruction(acq, reconParams)
slice_abs = abs.(image[:, :, 1])
p5 = plot_image(slice_abs)
```
And here's the .seq:
[spiralNormal.seq.txt](https://github.com/user-attachments/files/17611387/spiralNormal.seq.txt)

**Comment by cncastillo** [comment]
Hi! I believe the k-space FOV is too large for the sampling of the phantom, so you are getting aliasing artifacts. This happens because the spins are in a uniform rectangular grid, so they generate a repeating spectrum (separated in 1/dx).  You can see a spurious echo at the end! (red arrow)

![image](https://github.com/user-attachments/assets/43dec1a3-8710-4db0-9e5c-472f1ad69849)

To increase the phantom sampling you can use the `us` keyword:
```julia
obj = brain_phantom2D(; us=2) # Up-sample by a factor of 2
```
This produces:
![newplot](https://github.com/user-attachments/assets/a0ee8352-de37-4e9b-bd6e-b0943214ebed)

MRIReco, used for the reconstruction, uses a DCF based on the k-space samples, which could look slightly different from what you expect. If so, varying the DCF or reconstructing externally with the output raw data MRD would be advised.

BTW, KomaMRI contains KomaMRICore, KomaMRIFiles, and KomaMRIPlots. So `using KomaMRI` is the same as doing `using KomaMRICore, KomaMRIFiles, KomaMRIPlots`, the latter does not load the UI and recon code.

**Reply by nathanjbeaumont** [reply]
Thanks that looks more as expected.
I'm trying to recreate figure 7 here: https://onlinelibrary.wiley.com/doi/10.1002/mrm.25559
![image](https://github.com/user-attachments/assets/54c21fd5-45bf-4f38-8090-ebb3c71b7ffa)

The authors say
"All acquired spiral data were reconstructed using nonuniform fast Fourier transform with a separately measured spiral trajectory that would correct the gradient imperfection, such as from eddy current or gradient delays. Images from individual coils were combined using the adaptive coil combination method, and coil-combined images were normalized to coil sensitivity map to mitigate the image intensity variation due to the coil sensitivity"

Does KomaMRI have an option to do NUFFT with a DCF?

**Reply by cncastillo** [reply]
This is more of an MRIReco thing. I looked into it, and MRIReco already uses a NUFFT with a DCF if the trajectory is not cartesian. I think the only "problems" were:
- There could be a DC offset, as the DCF is not 0 for k=0 (Based on https://onlinelibrary.wiley.com/doi/epdf/10.1002/mrm.1910380117)
```julia
using MRIReco
density = samplingDensity(acq, reconParams[:reconSize])
density[1][1] = 0.0 # Fixes problem 1
```
- I believe the raw data is not pre-multiplied by sqrt(dcf) only the operators (the initial data b needs to be multiplied by it based on https://onlinelibrary.wiley.com/doi/epdf/10.1002/mrm.1241)
```julia
raw2 = copy(raw)
raw2.profiles[1].data .*= density[1] # Fixes problem 2
```
- The generated DCF is a little "noisy", I am not sure if a custom DCF can be used. Maybe @tknopp or @nHackel can help?

DCF from `MRIReco.samplingDensity` after modifications mentioned above:
![newplot](https://github.com/user-attachments/assets/f1394bb7-86ed-4527-911b-31753821d6bb)

Result (it should look sharper):
![newplot](https://github.com/user-attachments/assets/7f51d36a-7672-48c0-886a-103cc8d8866d)

Full code:
```julia
using KomaMRI, CUDA
obj = brain_phantom2D(; us=2)
seq = read_seq("spiralNormal.seq")
sys = Scanner() # default hardware definition
raw = simulate(obj, seq, sys)
acq = AcquisitionData(raw)
acq.traj[1].circular = false #This is to remove the circular mask
reconParams = Dict{Symbol,Any}(:reconSize=>(255, 255))
image = reconstruction(acq, reconParams)
slice_abs = abs.(image[:, :, 1])
p5 = plot_image(slice_abs)

# DCF modifications
using MRIReco
density = samplingDensity(acq, reconParams[:reconSize]) # Gives sqrt(dcf)
density[1][1] = 0.0 # Fixes problem 1
raw2 = copy(raw)
raw2.profiles[1].data .*= density[1]  # Fixes problem 2
acq2 = AcquisitionData(raw2)
acq2.traj[1].circular = false #This is to remove the circular mask
image2 = reconstruction(acq2, reconParams)
slice_abs2 = abs.(image2[:, :, 1])
p6 = plot_image(slice_abs2)
```

**Comment by aTrotier** [comment]
@cncastillo  the "noise" in the dcf is "normal" and won't impact a lot the image reconstruction.

The results seems ok to me, the spiral acquisition is especially subsampled (approximatively 256/8 turns = 16 times ?)  I guess it is done like that for a MRF like acquisition.

@nathanjbeaumont the image on the siemens scanner was reconstructed with an Ice pipeline ?

**Reply by nathanjbeaumont** [reply]
I don't know what an Ice pipeline is, but I looked into it and the Siemens reconstruction was actually done using a window of the first 24 interleaves of a 48 interleaf sequence, so actually disregard the phantom image. However I do think a single interleaf simulation should be able to produce an image like this:
![image](https://github.com/user-attachments/assets/1762417b-e3b4-465a-bf74-7bcc37d7f6c1)

**Reply by cncastillo** [reply]
I have used a similar recon for our MRF implementation and have gotten images like the ones you show, but the contrast changes through the MRF sequence. These are in vivo acquisitions, btw.

https://github.com/user-attachments/assets/1d33cf54-3240-4d78-98ab-cd63c8d4f7bb

For the first spiral readout, the image has a higher signal intensity, I would add a few more readouts. To reconstruct more images change Nz to the number of spiral readouts in the sequence definition.

Having said that, the reconstruction differences shouldn't affect the MRF quantification after dictionary matching as they are consistently different.

**Reply by cncastillo** [reply]
I forgot to say, but you can add readouts by scaling and rotating the single-shot acquisition.

```julia
seq = read_seq(...) # single-shot undersampled spiral
seq_mrf = Sequence()
for i in 1:Nreadouts
    seq_mrf += (rf_scaling[i] * cis(rf_phase[i])) * (rotz(spiral_angle[i]) * seq)
end
```

Gradients are rotated with a rotation matrix and RFs are scaled by a complex number.

**Reply by nathanjbeaumont** [reply]
Is each frame in the video you posted reconstruction of a single interleaf of a spiral? So the image quality varies drastically between TR indices? In this case [figure 7 here]( https://onlinelibrary.wiley.com/doi/10.1002/mrm.25559) is a cherry picked interleaf with one of the best looking reconstructions?

**Reply by cncastillo** [reply]
It could be a slightly different reconstruction, but yeah, the reconstructed interleafs can look quite different. The selected TR is probably the one that looked the best.
